solr/contrib/ltr/src/java/org/apache/solr/ltr/interleaving/TeamDraftInterleaving.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.ltr.interleaving;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Random;
+import java.util.Set;
+
+import org.apache.lucene.search.ScoreDoc;
+
+public class TeamDraftInterleaving implements Interleaving{
```

Comment:
8/n suggestions:
* class level javadocs re: the interleaving algorithm
* comments or javadocs re: any assumptions e.g.
  * must rerankedA.length and rerankedB.length match?
  * can rerankedA and rerankedB contain the same docs?
  * can rerankedA contain the same doc more than once?
  * can rerankedB contain the same doc more than once?
* consider guarding against array-index-out-of-bounds exceptions (even if they shouldn't happen if all assumptions are met)

```
indexA = updateIndex(interleavedResults,indexA,rerankedA);
if (indexA < rerankedA.length) {
  interleavedResults.add(rerankedA[indexA]);
  teamA.add(rerankedA[indexA].doc);
  indexA++;
}
```

Reply:
update the java docs, in regards to the index out of bounds, I think we should raise the exception instead of adding the condition:
if the assumptions are not met, there is a bug to be fixed, so we need interleaving to fail instead of silently operates wrong, am I missing anything?
Please take a look to my new commit, especially for the comments

Reply:
Very nice java docs and comments, thanks!

> ... I think we should raise the exception ...

Yes, you're right, an exception that will lead to a bug being identified and fixed rather than silent failure or (confusing for the user) "array out of bounds exception" would be better.